Fix duplicate download (#450)

jessebrennan opened the pull request:
Fixes race in file download where the file in the filestore could be overwritten, which would leave a dangling link to the file outside the filestore and thus a duplicate file.

hannes-ucsc commented on hca/dss/util/__init__.py:
```diff
@@ -51,3 +55,13 @@ def hardlink(source, link_name):
             raise
+
+
+@contextlib.contextmanager
```
This seems to conflate `FileExistsError`s raised by the body of surrounding `with` statement (which shouldn't be suppressed) with those raised by the CM itself. Looking at the code of atomicwrites it is not clear how we could fix this since it uses `contextlib.contextmanager` not the more direct approach of implementing `__enter__` and `__exit__` methods.

jessebrennan replied:
Wow, nice catch. I think I have a solution...

jessebrennan commented:
@hannes-ucsc tests were failing due to a flaky test. An issue has been created https://github.com/HumanCellAtlas/dcp-cli/issues/490

hannes-ucsc commented on hca/dss/__init__.py:
```diff
@@ -760,14 +759,14 @@ def _write_output_manifest(self):
         if 'file_path' not in fieldnames:
             fieldnames.append('file_path')
-        with atomic_write(output, overwrite=True, newline='') as f:
+        if os.path.isfile(output):
```
`isfile` returns True for symlinks as well. You may want to think through that case. I don't remember what `open(p,'w')` does 

Also, if you fail half-way through writing the manifest, you don't want the file to be half-written. Ideally, there should be no file or only a temporary file (not ideal). That's why `atomicwrite` may still be needed here.

hannes-ucsc commented on hca/dss/util/__init__.py:
```diff
@@ -51,3 +55,22 @@ def hardlink(source, link_name):
             raise
+
+
+class atomic_write:
```
```suggestion
class atomic_overwrite:
```

???

jessebrennan replied:
Much better! :joy:

hannes-ucsc commented on test/unit/test_dss_client.py:
```diff
@@ -82,10 +88,12 @@ def _fake_do_download_file_with_barrier(*args, **kwargs):
     fh.write(b'Here we write some stuff so that the fake download takes some time. '
              b'This helps ensure that multiple threads are writing at once and thus '
              b'allows us to test for race conditions.')
+    time.sleep(random.random())  # sleep for some small random amount of time
```
Comment is obvious. I think the string above explains the intent sufficiently.

jessebrennan replied:
Absolutely agree.

hannes-ucsc commented on test/unit/test_dss_client.py:
```diff
@@ -29,45 +31,49 @@ def _fake_download_file(*args, **kwargs):
-    for f in bundle_dict['files']:
-        f['bundle_uuid'] = kwargs['uuid']
-    yield {'bundle': bundle_dict}
+def _fake_paginate_factory(fake_hash=False):
```
```suggestion
def _make_fake_paginate(fake_hash=False):
```

???

jessebrennan replied:
Definitely better.

jessebrennan commented:
@hannes-ucsc to facilitate resolution of merge conflicts, I squashed previous fixups, however, I made sure to save the latest fixups in response to your last review so that it's easy to tell what I changed.

jessebrennan commented:
@hannes-ucsc in order to facilitate resolution of merge conflicts during rebasing, I had to squash the PR. I kept the review changes from your latest review as separate fixup! commits. I also added a couple new commits to address some broken tests.

hannes-ucsc commented on test/unit/util/__init__.py:
```diff
@@ -17,5 +23,34 @@ def test_dict_merge(self):
 
+
+class TestLinking(TmpDirTestCase):
+    """TmpDirTestCase will ensure any links / files we create are cleaned up"""
```
I don't like one-liner docstrings because I like the vertical space introduced by the two lines with just triple quotes on them. But there is no clear guide for this repo so I'll leave this to you.

jessebrennan replied:
Since this format of docstring is still PEP 8, I think it is acceptable to leave it as is.

hannes-ucsc commented:
Merge at will.
